### DISCUSSION TITLE
How can you count the number of tasks awaiting processing?

### DISCUSSION THREAD

**Discussion by mlissner** [celery/celery#8595] [Q&A]
I use celery to complete ETL tasks. To do so, I create a script that creates one task per item in our database, but our database contains millions of items, so I don't want to create millions of tasks all at once. Ideally, I'd like my ETL script to monitor celery and keep the workers fed at roughly the rate they consume tasks.

To do this, I need a way to figure out the number of tasks awaiting workers. I've been unable to figure this out for years, and there [are some solutions around](https://stackoverflow.com/questions/5544629/retrieve-list-of-tasks-in-a-queue-in-celery/43420719#43420719), but they fall short:

 - Using `inspect` will tell you the number of reserved tasks or the number of active tasks, but it won't tell you the number of "enquequed" tasks.
 - You can query your broker to get the length of the queue, but that also includes failed tasks.
 - You can grab the entire queue from your broker, deserialize, and inspect each task to see if it failed, but this doesn't perform.

As far as I can tell, there's no way to do this, but I'm flummoxed how that could be because workers know how to grab tasks from the queue that need processing. 

So my questions are:

1. Is there a way to identify the number of tasks that are ready for workers?

2. If that's not easy: How can workers distinguish which tasks are failed vs which need to be processed?

Thank you! 

---

 - This question has also been asked here: https://github.com/celery/celery/discussions/8116
 - And has many bad answers (including my own) here: https://stackoverflow.com/questions/5544629/retrieve-list-of-tasks-in-a-queue-in-celery
 - And more struggle here: https://stackoverflow.com/questions/61502706/celery-how-to-get-queue-size-in-a-reliable-and-testable-way

**Comment by Alekky09** [comment]
What do you mean by `enqueued` tasks?
Which broker do you use?

**Reply by mlissner** [reply]
Sorry for the delayed reply. By enqueued tasks, I mean tasks that are awaiting a worker to process them. 

I use redis as the broker?

**Comment by samhilmw** [comment]
Hello, If you got the answer can you share it over here @mlissner   
It will be helpful!

**Reply by mlissner** [reply]
It's been so long that I don't know if I was even asking the right question, but all the tools we have for celery are in https://github.com/freelawproject/courtlistener. I think they're in `cl.lib.celery_utils` or similar.
